x-pack/apm-server/profiling/ilm.go

```diff
@@ -0,0 +1,474 @@
+		return
+	}
+
+	policy := ilmPolicy{
+		cfg.ILMConfig.SizeInBytes,
+		cfg.ILMConfig.Age,
+	}
+	for _, idx := range keyValueIndices {
+		go func(index string) {
+			// Jitter and the atomicity of rollover/swap give enough guarantee that
+			// we avoid collisions between multiple instances of the collector.
+			// A more robust concurrency control is implemented in the rollover function.
+			randomizedTicker := time.NewTicker(AddJitter(cfg.ILMConfig.Interval, jitter))
+			runner := &optimisticAliasSwap{
+				client:       client,
+				currentAlias: index,
+				logger:       logger,
+			}
+			for {
+				// runILM is a blocking call, waiting for the ticker's tick or context expiration.
+				if err := runILM(ctx, randomizedTicker.C, maxRetries, runner, policy); err != nil {
+					if errors.Is(err, context.Canceled) {
+						return
+					}
+					logger.Warnf("Custom ILM strategy: %v", err)
+				}
+				randomizedTicker.Reset(AddJitter(cfg.ILMConfig.Interval, jitter))
+			}
+		}(idx)
+	}
+}
+
+// ILMRunner describes the behavior of a custom ILM strategy executor
+type ILMRunner interface {
+	configure() error
+	preFlight(policy ilmPolicy) error
+	execute(policy ilmPolicy) error
+}
+
+type optimisticAliasSwap struct {
+	client *es.Client
+	currentAlias,
+	nextIndex,
+	nextAlias,
+	currentIndex string
+
+	logger *logp.Logger
+}
+
+func (o *optimisticAliasSwap) configure() error {
+	return bootstrapILMLockDocument(o.client, o.currentAlias)
+}
+
+func (o *optimisticAliasSwap) preFlight(policy ilmPolicy) error {
+	var err error
+	o.currentIndex, err = indexFromAlias(o.client, o.currentAlias)
+	if err != nil {
+		return err
+	}
+
+	if !indexShouldBeRolledOver(o.logger, o.client, o.currentIndex, policy) {
+		return errPreFlightFailed{o.currentIndex}
+	}
+
+	o.nextAlias = nextIndex(o.currentAlias)
+	if o.nextIndex, err = indexFromAlias(o.client, o.nextAlias); err != nil {
+		return err
+	}
+
+	o.logger.With(logp.String("index", o.currentAlias)).
+		Debugf("Index status matching the ILM policy %s, possibly start roll-over", policy)
+
+	return nil
+}
+
+func (o *optimisticAliasSwap) execute(policy ilmPolicy) error {
+	ok, err := ilmAcquireLock(o.client, o.currentAlias, policy.age)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		// Early-exit, some other replica has acquired the lock
+		return nil
+	}
+	// Releasing the lock does not need to check on errors,
+	// because the lock acquire process is self-healing, based on the timerange
+	defer func() {
+		if err = ilmIndexLockDocument(o.client, o.currentAlias, "completed",
+			nil, nil); err != nil {
+			o.logger.Warnf("Failed to index document in lock index: %v", err)
+		}
+	}()
+	rolloverCounter.Inc()
+
+	// Create a new index: setting the right name will allow inheriting
+	// the configuration from the template
+	newIdx := fmt.Sprintf("%s-%s", o.currentAlias,
+		time.Now().UTC().Format("2006.02.01-15.04.05"))
+	if resp, err := o.client.Indices.Create(newIdx); checkESAPIError(http.StatusOK,
```

Comment:
I think you should always be closing resp.Body as long as err == nil. I didn't check through ES API calls, but all the ones I saw were using this pattern, and it only closes the body when the status code indicates an error.

Reply:
Are you referring to `checkESAPIError` in general or this particular call site?
What you mentioned is true for `http.Response.Body`, is it the same for `esapi.Response`?

Reply:
Yes, you are expected to close the response body for `esapi.Response`s too.

The `esapi.Response.String()` doc comment says it is intended for debugging. I guess it's fine, just pointing that out in case you missed it.

Reply:
NVM, just read the comment on top of it...

Reply:
So yeah, after re-reading `elastictransport.Client.Perform` implementation, I _guess_ it's safer to always close the body.
I'll apply these changes.